File: leetcode-js/74.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
    let left = 0;
    let right = matrix.length * matrix[0].length - 1;
    let index = Math.floor((left + right) / 2);

    const flatIdxToMat = (idx) => [Math.floor(idx / matrix[0].length), idx % matrix[0].length];

    while (left <= right) {
        let indexMat = flatIdxToMat(index);
        if (matrix[indexMat[0]][indexMat[1]] === target) return true;
        if (matrix[indexMat[0]][indexMat[1]] < target) left = index + 1;
        if (matrix[indexMat[0]][indexMat[1]] > target) right = index - 1;
        index = Math.floor((left + right) / 2);
    }
    return false;
};

// let matrix = [[1, 3, 5, 7], [10, 11, 16, 20], [23, 30, 34, 60]], target = 4;
let matrix = [[1]], target = 2;
console.log(searchMatrix(matrix, target));

File: leetcode-go/74.go
package main

import (
	"math"
)

func searchMatrix(matrix [][]int, target int) bool {
	left := 0
	right := len(matrix)*len(matrix[0]) - 1
	index := int(math.Floor(float64(left+right) / 2))
	for left <= right {
		indexX := int(math.Floor(float64(left+right)/2)) % len(matrix[0])
		indexY := int(math.Floor(float64(left+right) / 2 / float64(len(matrix[0]))))

		if matrix[indexY][indexX] == target {
			return true
		}
		if matrix[indexY][indexX] < target {
			left = index + 1
		}
		if matrix[indexY][indexX] > target {
			right = index - 1
		}
		index = int(math.Floor(float64(left+right) / 2))
	}
	return false
}
